### ISSUE TITLE
Job priorities

### ISSUE THREAD

**Issue by tobyink** [travis-ci/travis-ci#2566]
So if I have [twenty-three jobs](https://travis-ci.org/tobyink/p5-type-tiny/builds/30795844) kicked off by a git push, it would be nice if my `.travis.yml` file could tell Travis which ones I care the most about (so would like to be built first).

Something like:

```
matrix:
  include:
    - perl: "5.18"
      env: COVERAGE=1 EXTENDED_TESTING=1
      priority: 1
```

**Comment by mpedrera** [comment]
Sorry for the necrobump, but I also think this would be a very useful feature (I came here googling `travis job priority`).

In our case, we have a job that is slightly more fragile than the rest, which is the one that shoudl be run first. 

The rest of the jobs are also important and need to pass to consider the build successful, therefore `allow_failures` is discarded here. Some of them take a lot of time to complete, and this is where the `priority` feature would be handy.

So, workflow right now is:

```
1. Push changes
2. Some -long- jobs run before critical one -> Wait for them (can take hours)
3. Critical job runs -> Failure -> Go to 1
```

And this is what I'd like to have:

```
1. Push changes
2. Critical job runs, because it has higher priority -> Fails -> Go to 1
```

**Comment by joshk** [comment]
Thanks for the feedback.

Do you have a repo or build we can look at?

**Comment by toby-m** [comment]
I'm also interested in this. We have a number of low priority projects that are triggered fairly often as part of our localisation system (translation strings getting modified etc.) and when we have a wide reaching change they can create a large backlog that other work gets stuck behind.

**Comment by ezyang** [comment]
Also interested in this. Some jobs are more likely to fail; run those first!

**Comment by Lyricalpanda** [comment]
I see the use case too, but for me the clutter and complexity outweighs the benefit, so I'm closing this for now.

Excuse me, but what? Did I read this right? I fail to see how there is any clutter or complexity to a priority queue. A priority queue is a basic data structure that's been solved for awhile. A priority queue that has all jobs with a default is just a queue. Since you assign a priority as an override, there is no clutter here.

The use case is pretty straightforward. We have team A who pushes often, has a bigger team size, and wants to have automatic builds. They will be using the servers the most. We have Team B that is smaller in size and build times are 1/4 the size. If Team B is pushing less often they should have priority on the servers. It doesn't make sense for Team B to be held up by Team A on the times they push.

We're looking at travis-ci and want to use a priority queue for Team B. Our other alternatives is to make Team B wait behind Team A even though they're using significantly less % of the servers or to just reserve a server for Team B.

**Comment by ernado** [comment]
In my case Build Jobs are starting _AFTER_ some of the Allowed Failures ones, even with "the combination of allow_failures and fast_finish". Why Build Jobs have no priority by default?

**Comment by aames** [comment]
I'd also like to see a basic priority queue system as described by @Lyricalpanda - we have a similar situation and the described priority level would allow us maximum utilisation (i.e. we don't want to throttle Team A's builds) and would result in minimal queueing for Team B

**Comment by BanzaiMan** [comment]
For a simple allow/not allow job prioritizing (within a build matrix), see #1092.

**Comment by wagenet** [comment]
I would like to be able to prioritize different repos.

**Comment by aronisstav** [comment]
I want this too. Running my testsuite with code coverage reporting on the latest supported version is a slower job, cannot be easily 'split down' and should not block the execution of running the testsuite in older supported versions. I can probably cut time in half if I have this.

I tried stages, but this blocks the older versions. (`cover` job is the code coverage version, which runs `tests` and `tests-real`)
https://travis-ci.org/parapluu/Concuerror/builds/286097504

I'd try a plain include, but then the `cover` job will be scheduled last in the matrix.

**Comment by willlarche** [comment]
I would like this too. If we have 10 contributors all with PRs in the queue and then we need to suddenly test a hotfix release, we have to cancel all of them (and each has 4) in order to get the hotfix tested asap.

**Comment by pmontrasio** [comment]
I also want this. A git push on a branch that triggers tests+build+deploy is much more valuable than a git push on a development branch that triggers only tests. The one for the deploy must start immediately. It's not even a matter of priority: there should be two different queues. test-only push go to the normal queue, deploy-push go to their own queue.
